File: Searching_Algorithms/number-of-occurences.cpp
//{ Driver Code Starts
#include<bits/stdc++.h>

using namespace std;


// } Driver Code Ends
//User function template for C++
class Solution{
public:	
	/* if x is present in arr[] then returns the count
		of occurrences of x, otherwise returns 0. */
	int solve(int l,int h,int arr[],int n,int x){
        if(l > h){
            return 0;
         }
        int mid=l+(h-l)/2;
        if(arr[mid] == x){
            return 1+solve(l,mid-1,arr,n,x)+solve(mid+1,h,arr,n,x);
        }
        else if(arr[mid] > x){
            return solve(l,mid-1,arr,n,x);
        }
        else{
            return solve(mid+1,h,arr,n,x);
        }
    }
 
    int count(int arr[], int n, int x) {
        return solve(0,n-1,arr,n,x);
    }
};

//{ Driver Code Starts.

int main() {
    int t;
    cin >> t;
    while (t--) {
        int n, x;
        cin >> n >> x;
        int arr[n];
        for (int i = 0; i < n; i++) {
            cin >> arr[i];
        }
        Solution ob;
        auto ans = ob.count(arr, n, x);
        cout << ans << "\n";
    }
    return 0;
}

// } Driver Code Ends

File: Basic-DSA-Quest/number-of-occurences.java
//{ Driver Code Starts
// Initial Template for Java

import java.io.*;
import java.util.*;

class GFG {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int tc = Integer.parseInt(br.readLine().trim());

        while (tc-- > 0) {

            String[] str = br.readLine().trim().split(" ");
            int[] a = new int[str.length];
            for (int i = 0; i < str.length; i++) {
                a[i] = Integer.parseInt(str[i]);
            }
            String[] nk = br.readLine().trim().split(" ");
            int k = Integer.parseInt(nk[0]);
            Solution sln = new Solution();
            int ans = sln.countFreq(a, k);

            System.out.println(ans);
            System.out.println("~");
        }
    }
}
// } Driver Code Ends


class Solution {
    int countFreq(int[] arr, int target) {
        // code here
       int l  = 0;
       int n =  arr.length;
       int h = n-1;
       
         int count = 0;
    
    while (l <= h) {
        int mid = l + (h - l) / 2;
        
        if (arr[mid] == target) {
            // Count the occurrences of `x` on the left side
            int left = mid - 1;
            while (left >= l && arr[left] == target) {
                count++;
                left--;
            }
            
            // Count the occurrences of `x` on the right side
            int right = mid + 1;
            while (right <= h && arr[right] == target) {
                count++;
                right++;
            }
            
            // Include the middle element itself
            count++;
            
            // Now return the total count
            break;
        }
        else if (arr[mid] > target) {
            h = mid - 1;
        }
        else {
            l = mid + 1;
        }
    }
    
    return count;
    }
}
